pkg/providers/cloudstack/cloudstack.go

```diff
@@ -576,10 +576,29 @@ func (p *cloudstackProvider) needsNewKubeadmConfigTemplate(workerNodeGroupConfig
 	return true, nil
 }
 
+// AnyImmutableFieldChanged Used by EKS-A controller and CLI upgrade workflow to compare generated CSDC/CSMC's from
+// CAPC resources in fetcher.go with those already on the cluster when deciding whether or not to generate and apply
+// new CloudStackMachineTemplates
 func AnyImmutableFieldChanged(oldCsdc, newCsdc *v1alpha1.CloudStackDatacenterConfig, oldCsmc, newCsmc *v1alpha1.CloudStackMachineConfig) bool {
-	if !oldCsdc.Spec.Equal(&newCsdc.Spec) {
+	if len(newCsdc.Spec.AvailabilityZones) != len(oldCsdc.Spec.AvailabilityZones) {
 		return true
 	}
+	oAzsMap := map[string]v1alpha1.CloudStackAvailabilityZone{}
```

Comment:
I thought we already had a method in the api package to make this comparison.

Reply:
Unfortunately, that was the whole problem. When we make this comparison in this method, we actually compare the existing CSDC with the one that was generated from a CloudStackCluster resource in fetcher.go. That generated CSDC is missing an attribute - ManagementApiEndpoint (because it’s not easily available in the CloudStackCluster), so we can’t do a full comparison with .Equal()